### PR TITLE
Alpine Dockerfile with Lua

### PR DESCRIPTION
**Pull request by NinoSkopac**
For running Lua auth in containers. 
Makes installing Lua modules easy via `luarocks`.

### CODE REVIEW

File: g3proxy/docker/lua.alpine.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+RUN luarocks install dkjson
+
+FROM rust:alpine AS builder
+WORKDIR /usr/src/g3
+COPY . .
+RUN apk add --no-cache musl-dev cmake capnproto-dev openssl-dev openssl-libs-static c-ares-dev c-ares-static lua5.4-dev pkgconfig
+ENV PKG_CONFIG_PATH=/usr/lib/pkgconfig
+RUN cargo build --profile release-lto \
+ --no-default-features --features quic,c-ares,hickory,lua54 \
+ -p g3proxy -p g3proxy-ctl
```
Comment (zh-jq-b) [inline g3proxy/docker/lua.alpine.Dockerfile]: You may also want to compile in g3proxy-lua, which can be used to verify the call result of the script

Comment (kapunakap) [reply g3proxy/docker/lua.alpine.Dockerfile]: Append -p g3proxy-lua to line 18?

**Comment by zh-jq-b** [comment]
Is it possible to use dynamic linking? I see nginx-mod-http-lua is available in alpine, which seems to be dynamic linked.
We may use the official packaged rustc in alpine to do this? https://github.com/rust-lang/compiler-team/issues/422#issuecomment-2016379564

File: g3proxy/docker/lua.alpine.Dockerfile (hunk 1 above)
Comment (zh-jq-b) [reply g3proxy/docker/lua.alpine.Dockerfile]: yes

**Comment by zh-jq-b** [comment]
`RUSTFLAGS="-Ctarget-feature=-crt-static"` this can be used to disable static linking when using official rust:alpine containers

**Comment by NinoSkopac** [comment]
I'll look into this a bit more tonight. I forgot to note this is my first time using Lua. Actually, I'll have to move away from Lua to Python because Lua doesn't have first-party Google Cloud support (no SDKs). After I do that, I'll contribute the Python Dockerfile as well.

File: g3proxy/docker/lua.alpine.Dockerfile (hunk 1 above)
Comment (kapunakap) [reply g3proxy/docker/lua.alpine.Dockerfile]: okay you got it, I'll add that tonight after testing it.

**Comment by zh-jq** [comment]
you can support both lua and python features in the same docker file. It has to be dynamic linked to be able to call c based lua/python libs I guess.

**Comment by NinoSkopac** [comment]
I looked into the comment you posted, which led me to learn a bit about musl, and frankly I've no idea about all this dynamic and static linking. I can't even conceptualize it, really.. It's ok. I'll add Python, and if you wanna take it further, that's great.

File: g3proxy/docker/lua.alpine.Dockerfile (hunk 1 above)
Comment (kapunakap) [reply g3proxy/docker/lua.alpine.Dockerfile]: I've added the bit of code but it doesn't work.

Inside the container, after rebuilding and running:
```bash
ls /usr/bin/g3proxy
g3proxy      g3proxy-ctl
```

There's no g3proxy-lua binary.

**Comment by zh-jq-b** [comment]
I will switch the default alpine build to use dynamic first, then you can sync the changes to this script-enabled docker build file.

File: g3proxy/docker/lua.alpine.Dockerfile (hunk 1 above)
Comment (zh-jq-b) [reply g3proxy/docker/lua.alpine.Dockerfile]: COPY is also needed in the final build step.

**Comment by zh-jq-b** [comment]
6f21d548367005216f551b4d684e4879e0e016a9 updated to use dynamic linking

**Comment by NinoSkopac** [comment]
I added the missing COPY g3proxy-lua

File: g3proxy/docker/lua.alpine.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+# Copy Lua binaries and modules from the lua stage
+COPY --from=lua /usr/local/bin/lua /usr/local/bin/lua
+COPY --from=lua /usr/local/bin/luarocks /usr/local/bin/luarocks
+COPY --from=lua /usr/local/lib/lua /usr/local/lib/lua
+COPY --from=lua /usr/local/share/lua /usr/local/share/lua
+COPY --from=lua /usr/local/lib/luarocks /usr/local/lib/luarocks
+# Copy the compiled binaries from the builder stage
+COPY --from=builder /usr/src/g3/target/release-lto/g3proxy /usr/bin/g3proxy
+COPY --from=builder /usr/src/g3/target/release-lto/g3proxy-ctl /usr/bin/g3proxy-ctl
+COPY --from=builder /usr/src/g3/target/release-lto/g3proxy-ctl /usr/bin/g3proxy-lua
```
Comment (zh-jq) [inline g3proxy/docker/lua.alpine.Dockerfile]: the src file is still g3proxy-ctl

File: g3proxy/docker/lua.alpine.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+# beware: lua's c modules (e.g. cjson)
+# need to be statically linked
+# and won't work when installing in
+# alpine via luarocks. 
+RUN luarocks install dkjson
+
+FROM rust:alpine AS builder
+WORKDIR /usr/src/g3
+COPY . .
+RUN apk add --no-cache musl-dev cmake capnproto-dev openssl-dev openssl-libs-static c-ares-dev c-ares-static lua5.4-dev pkgconfig
```
Comment (zh-jq) [inline g3proxy/docker/lua.alpine.Dockerfile]: openssl-libs-static and c-ares-static can be removed

**Review by zh-jq — CHANGES_REQUESTED** [review]

File: g3proxy/docker/lua.alpine.Dockerfile (hunk 2 above)
Comment (kapunakap) [reply g3proxy/docker/lua.alpine.Dockerfile]: my bad!

File: g3proxy/docker/lua.alpine.Dockerfile (hunk 3 above)
Comment (kapunakap) [reply g3proxy/docker/lua.alpine.Dockerfile]: my bad, again

**Comment by NinoSkopac** [comment]
Done both, error is during build, during `cargo build` step:

```
 => => #   = note: /usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/../../../../aarch64-alpine-linux-musl/bin/ld: cannot find -lssl: No such file or directory                                                                                                                                                                                                                                                                                                                                 
 => => #           /usr/lib/gcc/aarch64-alpine-linux-musl/13.2.1/../../../../aarch64-alpine-linux-musl/bin/ld: cannot find -lcrypto: No such file or directory                                                                                                                                                                                                                                                                                                                              
 => => #           collect2: error: ld returned 1 exit status                                                                                                                                                                                                                                                                                                                                                                                                                               
 => => # error: could not compile `g3proxy-ctl` (bin "g3proxy-ctl") due to 1 previous error                                                                                                                                                                                                                                                                                                                                                                                                 
 => => # warning: build failed, waiting for other jobs to finish...           
```

**Comment by zh-jq-b** [comment]
It seems that package `libssl3` didn't get installed. I can't reproduce this on x86-64, so I guess this may be an Alpine bug?

File: g3proxy/docker/lua.alpine.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,33 @@
+FROM nickblah/lua:5.4-luarocks-alpine AS lua
```
Comment (zh-jq-b) [inline g3proxy/docker/lua.alpine.Dockerfile:1]: For the luarocks part, how about install `apk add luarocks-5.4` in the final stage, and then create a symbolic link from /usr/bin/luarocks-5.4 to /usr/bin/luarocks?

Comment (zh-jq-b) [reply g3proxy/docker/lua.alpine.Dockerfile:1]: And also lua5.4-socket, lua5.4-pgmoon (for posgresql),  lua5.4-sql-mysql, lua5.4-sql-postgres are packaged in alpine community channel. I'm more prefer to use apk to install lua modules.

File: g3proxy/docker/lua.alpine.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+# need to be statically linked
+# and won't work when installing in
+# alpine via luarocks. 
+RUN luarocks install dkjson
+
+FROM rust:alpine AS builder
+WORKDIR /usr/src/g3
+COPY . .
+RUN apk add --no-cache musl-dev cmake capnproto-dev openssl-dev c-ares-dev lua5.4-dev pkgconfig
+ENV PKG_CONFIG_PATH=/usr/lib/pkgconfig
```
Comment (zh-jq-b) [inline g3proxy/docker/lua.alpine.Dockerfile:15]: You missed this line for dynamic linking:
```
ENV RUSTFLAGS="-Ctarget-feature=-crt-static"
```

Comment (kapunakap) [reply g3proxy/docker/lua.alpine.Dockerfile:15]: apologies, in my mind i was merging in your changes :D

fixing

**Comment by NinoSkopac** [comment]
It builds, but here's what happens next.

```
docker run --rm -p 11080:11080 -p 13128:13128 -v ./g3proxy/:/etc/g3proxy/ g3proxy -c /etc/g3proxy/config.yaml
Error loading shared library liblua-5.4.so.0: No such file or directory (needed by /usr/bin/g3proxy)
Error loading shared library libcares.so.2: No such file or directory (needed by /usr/bin/g3proxy)
Error loading shared library libgcc_s.so.1: No such file or directory (needed by /usr/bin/g3proxy)
Error relocating /usr/bin/g3proxy: lua_getmetatable: symbol not found
Error relocating /usr/bin/g3proxy: ares_process_fd: symbol not found
Error relocating /usr/bin/g3proxy: ares_set_local_ip6: symbol not found
Error relocating /usr/bin/g3proxy: lua_tointegerx: symbol not found
Error relocating /usr/bin/g3proxy: lua_setmetatable: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_GetRegionStart: symbol not found
Error relocating /usr/bin/g3proxy: lua_rawgetp: symbol not found
Error relocating /usr/bin/g3proxy: ares_init_options: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_SetGR: symbol not found
Error relocating /usr/bin/g3proxy: luaL_requiref: symbol not found
Error relocating /usr/bin/g3proxy: lua_atpanic: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_GetDataRelBase: symbol not found
Error relocating /usr/bin/g3proxy: lua_setupvalue: symbol not found
Error relocating /usr/bin/g3proxy: ares_library_init: symbol not found
Error relocating /usr/bin/g3proxy: ares_destroy: symbol not found
Error relocating /usr/bin/g3proxy: lua_getallocf: symbol not found
Error relocating /usr/bin/g3proxy: ares_strerror: symbol not found
Error relocating /usr/bin/g3proxy: luaL_checkstack: symbol not found
Error relocating /usr/bin/g3proxy: luaL_newstate: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_DeleteException: symbol not found
Error relocating /usr/bin/g3proxy: lua_createtable: symbol not found
Error relocating /usr/bin/g3proxy: luaL_openlibs: symbol not found
Error relocating /usr/bin/g3proxy: lua_pushnumber: symbol not found
Error relocating /usr/bin/g3proxy: lua_pcallk: symbol not found
Error relocating /usr/bin/g3proxy: lua_pushinteger: symbol not found
Error relocating /usr/bin/g3proxy: lua_xmove: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_GetLanguageSpecificData: symbol not found
Error relocating /usr/bin/g3proxy: lua_rawgeti: symbol not found
Error relocating /usr/bin/g3proxy: lua_newstate: symbol not found
Error relocating /usr/bin/g3proxy: lua_touserdata: symbol not found
Error relocating /usr/bin/g3proxy: ares_library_cleanup: symbol not found
Error relocating /usr/bin/g3proxy: lua_pushlightuserdata: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_RaiseException: symbol not found
Error relocating /usr/bin/g3proxy: lua_tothread: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_FindEnclosingFunction: symbol not found
Error relocating /usr/bin/g3proxy: lua_checkstack: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_GetIP: symbol not found
Error relocating /usr/bin/g3proxy: lua_isinteger: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_Backtrace: symbol not found
Error relocating /usr/bin/g3proxy: lua_error: symbol not found
Error relocating /usr/bin/g3proxy: ares_set_local_ip4: symbol not found
Error relocating /usr/bin/g3proxy: luaL_loadstring: symbol not found
Error relocating /usr/bin/g3proxy: lua_settop: symbol not found
Error relocating /usr/bin/g3proxy: lua_rawequal: symbol not found
Error relocating /usr/bin/g3proxy: lua_toboolean: symbol not found
Error relocating /usr/bin/g3proxy: ares_parse_aaaa_reply: symbol not found
Error relocating /usr/bin/g3proxy: lua_pushvalue: symbol not found
Error relocating /usr/bin/g3proxy: lua_tonumberx: symbol not found
Error relocating /usr/bin/g3proxy: ares_parse_a_reply: symbol not found
Error relocating /usr/bin/g3proxy: luaL_loadbufferx: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_GetIPInfo: symbol not found
Error relocating /usr/bin/g3proxy: lua_pushboolean: symbol not found
Error relocating /usr/bin/g3proxy: lua_type: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_GetCFA: symbol not found
Error relocating /usr/bin/g3proxy: lua_gettop: symbol not found
Error relocating /usr/bin/g3proxy: lua_close: symbol not found
Error relocating /usr/bin/g3proxy: lua_tolstring: symbol not found
Error relocating /usr/bin/g3proxy: lua_gc: symbol not found
Error relocating /usr/bin/g3proxy: lua_rawset: symbol not found
Error relocating /usr/bin/g3proxy: ares_query: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_GetTextRelBase: symbol not found
Error relocating /usr/bin/g3proxy: lua_rotate: symbol not found
Error relocating /usr/bin/g3proxy: ares_set_servers_ports_csv: symbol not found
Error relocating /usr/bin/g3proxy: lua_gettable: symbol not found
Error relocating /usr/bin/g3proxy: luaL_traceback: symbol not found
Error relocating /usr/bin/g3proxy: lua_rawsetp: symbol not found
Error relocating /usr/bin/g3proxy: luaL_tolstring: symbol not found
Error relocating /usr/bin/g3proxy: lua_topointer: symbol not found
Error relocating /usr/bin/g3proxy: lua_pushcclosure: symbol not found
Error relocating /usr/bin/g3proxy: lua_pushlstring: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_Resume: symbol not found
Error relocating /usr/bin/g3proxy: lua_copy: symbol not found
Error relocating /usr/bin/g3proxy: _Unwind_SetIP: symbol not found
Error relocating /usr/bin/g3proxy: lua_pushnil: symbol not found
Error relocating /usr/bin/g3proxy: luaL_ref: symbol not found
Error relocating /usr/bin/g3proxy: lua_newthread: symbol not found
Error relocating /usr/bin/g3proxy: lua_newuserdatauv: symbol not found
Error relocating /usr/bin/g3proxy: lua_rawget: symbol not found
Error relocating /usr/bin/g3proxy: lua_atpanic: symbol not found
Error relocating /usr/bin/g3proxy: luaL_loadstring: symbol not found
Error relocating /usr/bin/g3proxy: luaL_openlibs: symbol not found
Error relocating /usr/bin/g3proxy: luaopen_table: symbol not found
Error relocating /usr/bin/g3proxy: luaopen_io: symbol not found
Error relocating /usr/bin/g3proxy: luaopen_string: symbol not found
Error relocating /usr/bin/g3proxy: lua_tocfunction: symbol not found
Error relocating /usr/bin/g3proxy: luaopen_math: symbol not found
Error relocating /usr/bin/g3proxy: lua_setglobal: symbol not found
Error relocating /usr/bin/g3proxy: luaopen_utf8: symbol not found
Error relocating /usr/bin/g3proxy: luaopen_coroutine: symbol not found
Error relocating /usr/bin/g3proxy: luaopen_package: symbol not found
Error relocating /usr/bin/g3proxy: luaopen_debug: symbol not found
Error relocating /usr/bin/g3proxy: luaopen_base: symbol not found
Error relocating /usr/bin/g3proxy: lua_isuserdata: symbol not found
Error relocating /usr/bin/g3proxy: luaL_setfuncs: symbol not found
Error relocating /usr/bin/g3proxy: luaopen_os: symbol not found
Error relocating /usr/bin/g3proxy: lua_getglobal: symbol not found
```

**Comment by zh-jq-b** [comment]
You need to install the missing dynamic libs before COPY.
`RUN apk add --no-cache libgcc c-ares lua5.4-libs`

**Comment by NinoSkopac** [comment]
Got it, I'll be pushing that tonight after testing.

**Comment by NinoSkopac** [comment]
runs after https://github.com/bytedance/g3/pull/292/commits/736e36dfb3e0b36b86afd522d9611a392e70ed49 as per your comment

**Review by zh-jq — APPROVED** [review]

File: g3proxy/docker/lua.alpine.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,35 @@
+FROM nickblah/lua:5.4-luarocks-alpine AS lua
+RUN apk add --no-cache build-base
+# install lua modules here.
+# dkjson is just an example.
+# beware: lua's c modules (e.g. cjson)
+# need to be statically linked
+# and won't work when installing in
+# alpine via luarocks.
```
Comment (kapunakap) [inline g3proxy/docker/lua.alpine.Dockerfile:8]: this comment is no longer valid. fixing.

**Pull request merged**